staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go

```diff
@@ -0,0 +1,1396 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer tearDown()
+
+	group := myCRDV1Beta1.Group
+	version := myCRDV1Beta1.Version
+	resource := myCRDV1Beta1.Resource
+	kind := fakeRESTMapper[myCRDV1Beta1]
+
+	myCRD := &apiextensionsv1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{Name: resource + "." + group},
+		Spec: apiextensionsv1.CustomResourceDefinitionSpec{
+			Group: group,
+			Versions: []apiextensionsv1.CustomResourceDefinitionVersion{{
+				Name:    version,
+				Served:  true,
+				Storage: true,
+				Schema: &apiextensionsv1.CustomResourceValidation{
+					OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+						Type: "object",
+						Properties: map[string]apiextensionsv1.JSONSchemaProps{
+							"content": {
+								Type: "object",
+								AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+									Schema: &apiextensionsv1.JSONSchemaProps{
+										Type: "string",
+									},
+								},
+							},
+							"num": {
+								Type: "object",
+								AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+									Schema: &apiextensionsv1.JSONSchemaProps{
+										Type: "integer",
+									},
+								},
+							},
+						},
+					},
+				},
+			}},
+			Names: apiextensionsv1.CustomResourceDefinitionNames{
+				Plural:   resource,
+				Kind:     kind,
+				ListKind: kind + "List",
+			},
+			Scope: apiextensionsv1.NamespaceScoped,
+		},
+	}
+
+	_, err = fixtures.CreateNewV1CustomResourceDefinition(myCRD, apiExtensionClient, dynamicClient)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, c := range cases {
+		if c.Disabled {
+			continue
+		}
+
+		t.Run(c.Name, func(t *testing.T) {
+			ctx := &ratchetingTestContext{
+				T:                   t,
+				DynamicClient:       dynamicClient,
+				APIExtensionsClient: apiExtensionClient,
+			}
+
+			for i, op := range c.Operations {
+				if err := op(ctx); err != nil {
+					t.Fatalf("failed operation %v: %v", i, err)
+				}
+			}
+
+			// Reset resources
+			err := ctx.DynamicClient.Resource(myCRDV1Beta1).Namespace("default").DeleteCollection(context.TODO(), metav1.DeleteOptions{}, metav1.ListOptions{})
+			if err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+var myCRDV1Beta1 schema.GroupVersionResource = schema.GroupVersionResource{
+	Group:    "mygroup.example.com",
+	Version:  "v1beta1",
+	Resource: "mycrds",
+}
+
+var myCRDInstanceName string = "mycrdinstance"
+
+func TestRatchetingFunctionality(t *testing.T) {
+	cases := []ratchetingTestCase{
+		{
+			Name: "Minimum Maximum",
+			Operations: []ratchetingTestOperation{
+				updateMyCRDV1Beta1Schema(&apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"hasMinimum":           *numberSchema,
+						"hasMaximum":           *numberSchema,
+						"hasMinimumAndMaximum": *numberSchema,
+					},
+				}),
+				// Create an object that complies with the schema
```

Comment:
it helps a lot during debugging to turn these comments into `t.Log`. Rule of thumb: the logs should tell a story.

Reply:
converted the comments into logs